v-next/hardhat/src/internal/cli/main.ts

```diff
@@ -265,3 +269,192 @@ function getTaskFromCliArguments(
+        },
+      );
+    }
+
+    usedCliArguments[i] = true;
+
+    if (paramInfo.parameterType === ParameterType.BOOLEAN) {
+      if (
+        usedCliArguments[i + 1] === false &&
+        ["true", "false"].includes(cliArguments[i + 1])
+      ) {
+        // The flag could be follow by the boolean value
+        taskArguments[paramName] = formatParameterValue(
+          cliArguments[i + 1],
+          ParameterType.BOOLEAN,
+          paramName,
+        );
+
+        usedCliArguments[i + 1] = true;
+        continue;
+      }
+
+      // If the flag is not followed by a boolean value, then the flag itself means that the value is true
+      taskArguments[paramName] = true;
+      continue;
+    }
+
+    // The value immediately following a named parameter (if the parameter does not behave as a flag)
+    // is the parameter's value; otherwise, it's an error
+    if (usedCliArguments[i + 1] === false) {
+      i++;
+
+      taskArguments[paramName] = formatParameterValue(
+        cliArguments[i],
+        paramInfo.parameterType,
+        paramName,
+      );
+
+      usedCliArguments[i] = true;
+
+      continue;
+    }
+
+    throw new HardhatError(
+      HardhatError.ERRORS.ARGUMENTS.MISSING_VALUE_FOR_NAMED_PARAMETER,
+      {
+        paramName: arg,
+      },
+    );
+  }
+}
+
+function parsePositionalAndVariadicParameters(
+  cliArguments: string[],
+  usedCliArguments: boolean[],
+  task: Task,
+  taskArguments: Record<string, any>,
+) {
+  let paramI = 0;
+
+  for (let i = 0; i < cliArguments.length; i++) {
+    if (usedCliArguments[i] === true) {
+      continue;
+    }
+
+    if (cliArguments[i] === "--") {
+      // A standalone '--' is ok because it is used to separate CLI tool arguments from script arguments,
+      // ensuring the tool passes subsequent options directly to the script.
+      usedCliArguments[i] = true;
+      continue;
+    }
+
+    const paramInfo = task.positionalParameters[paramI];
+
+    if (paramInfo === undefined) {
+      continue;
+    }
+
+    usedCliArguments[i] = true;
+
+    const formattedValue = formatParameterValue(
+      cliArguments[i],
+      paramInfo.parameterType,
+      paramInfo.name,
+    );
+
+    if (paramInfo.isVariadic === false) {
+      taskArguments[task.positionalParameters[paramI++].name] = formattedValue;
+      continue;
+    }
+
+    // Handle variadic parameters
+    taskArguments[paramInfo.name] = taskArguments[paramInfo.name] ?? [];
+    taskArguments[paramInfo.name].push(formattedValue);
+  }
+}
+
+function kebabToCamelCase(str: string) {
+  return str.replace(/-./g, (match) => match.charAt(1).toUpperCase());
+}
+
+function formatParameterValue(
+  value: string,
+  type: ParameterType,
+  name: string,
+): any {
+  switch (type) {
+    case ParameterType.STRING:
+    case ParameterType.FILE:
+      return value;
+    case ParameterType.INT:
+    case ParameterType.FLOAT:
+      return Number(value);
+    case ParameterType.BIGINT:
+      return BigInt(value);
```

Comment:
These should also handle errors.

Also, in v2 we were accepting these strings:

* int     
```js
const decimalPattern = /^\d+(?:[eE]\d+)?$/;
const hexPattern = /^0[xX][\dABCDEabcde]+$/;
```

* bigint
```
const decimalPattern = /^\d+(?:n)?$/;
const hexPattern = /^0[xX][\dABCDEabcde]+$/;
```

* float
```
const decimalPattern = /^(?:\d+(?:\.\d*)?|\.\d+)(?:[eE]\d+)?$/;
const hexPattern = /^0[xX][\dABCDEabcde]+$/;
```

Reply:
Should we consider to create a function for these operations in hh-utils? I could add the functions there and then use them here. WDYT?

Reply:
I wouldn't, as these have user-facing behavior that can accidentally be changed when/if we change the utils package. I think, in this case, locality is more important than reusability.

Reply:
Done it [this commit](https://github.com/NomicFoundation/hardhat/pull/5300/commits/bab2ad3654f68b050182e8e424254774d9800078) but follow up question: for the FILE type should we return the path passed by the user or the absolute path? Currently I'm returning the absolute path.

Also, the type validation modified a function from sync to async, so in this commit you'll see quite a lot of code changed because a lot of functions had to be converted into async to wait for promises

Reply:
After our offline sync we decided to handle FILE types as normal strings. So I will have to change the validation for the FILE type and undo the changes related to the async functions because the async function was the one used to read the folder but now it's no longer needed